### PR TITLE
US5121: YES! Auth + Guest Giving is Finally Ready!

### PR DESCRIPTION
**Pull request by samueljmello**

### CODE REVIEW

File: src/app/authentication/authentication.component.spec.ts [hunk 1]
```diff
@@ -1,54 +1,192 @@
...
+      });
+    });
+
+    describe('when invalid credentials are submitted', () => {
+      beforeEach(() => {
+        setForm('bad@bad.com', 'reallynotgood');
+        (<jasmine.Spy>loginService.login).and.returnValue(Observable.throw({}));
+      });
+
+      it('#adv does not get called', () => {
```
Comment (vchtelmakh) [inline src/app/authentication/authentication.component.spec.ts]: Convention is that it statements start with 'should', e.g.  'should not call adv'. Not important, no need to change.

File: src/app/billing/billing.component.ts [hunk 2]
```diff
@@ -231,13 +169,48 @@ export class BillingComponent implements OnInit {
...
-    this.stateManagerService.is_loading = false;
+  private handleDonorError(errResponse: any, isCC = false): boolean {
+    if ( isCC === true ) {
+      this.setCCSubmitted(false);
+    } else {
+      this.setACHSubmitted(false);
+    }
+    this.state.setLoading(false);
+    if (errResponse.status === 400 || errResponse.status === 500) {
+      this.gift.systemException = true;
```
Comment (vchtelmakh) [inline src/app/billing/billing.component.ts]: Need to this.setACHSubmitted(false); and/or turn off spinner on failed call?

File: src/app/billing/billing.component.ts [hunk 3]
```diff
@@ -54,175 +55,112 @@ export class BillingComponent implements OnInit {
...
     return false;
   }
 
-  ccNext() {
-    this.ccSubmitted = true;
-    this.gift.stripeException = false;
-    this.gift.systemException = false;
+  public ccNext() {
+
+    this.setCCSubmitted(true);
```
Comment (vchtelmakh) [inline src/app/billing/billing.component.ts]: Is this also not used?

File: src/app/services/login.service.ts [hunk 4]
```diff
@@ -3,6 +3,9 @@ import { Response } from '@angular/http';
 import { Observable } from 'rxjs/Rx';
 import { HttpClientService } from './http-client.service';
 
+import 'rxjs/add/operator/catch';
```
Comment (vchtelmakh) [inline src/app/services/login.service.ts]: Don't need to import these separately if you import import { Observable } from 'rxjs/Rx';.

File: src/app/summary/summary.component.ts [hunk 5]
```diff
@@ -20,46 +20,98 @@ export function _window(): Window {
...
+    if (this.isSubmitInProgress) {
+      return;
+    }
+
+    this.beginProcessing();
+
+    if (this.gift.isOneTimeGift()) {
+      let pymt_type = this.gift.paymentType === 'ach' ? 'bank' : 'cc';
+      let donationDetails = new PaymentCallBody(this.gift.fund.ProgramId.toString(), this.gift.amount,
+                                                pymt_type, 'DONATION', this.gift.invoiceId );
```
Comment (richardbarker) [inline src/app/summary/summary.component.ts]: If this is a donation, what are we setting invoiceId to? Should we even set it?

Comment (vchtelmakh) [inline src/app/summary/summary.component.ts]: @richardbarker iirc it just goes as default, which is 0, and doesn't get used by the API.

File: src/app/billing/billing.component.ts [hunk 6]
```diff
@@ -54,175 +55,112 @@ export class BillingComponent implements OnInit {
...
-    }
-    return ret;
-  }
+  public achNext() {
 
-  achNext() {
-    this.achSubmitted = true;
-    this.gift.stripeException = false;
-    this.gift.systemException = false;
+    this.setACHSubmitted(true);
```
Comment (samueljmello) [inline src/app/billing/billing.component.ts]: this controls whether or not the form submitted class for ACH is on or off (Basically).

File: src/app/summary/summary.component.ts (hunk 5 above)
Comment (samueljmello) [inline src/app/summary/summary.component.ts]: it was more of a copy/paste result from using the payment stuff. We can just safely force 0.

File: src/app/authentication/authentication.component.html [hunk 7]
```diff
@@ -8,38 +12,40 @@
...
+      <p>Password <span [innerHTML]="formatErrorMessage(form.controls['password'].errors)"></span></p>
+    </div>
 
-      <div class="form__helpers--half">
+    <div class="row form__helpers">
+      <div class="form__helpers--full">
         <p class="pull-right text-left text-block">
-          <a class="text-block__link" routerLink="/password">Forgot password?</a><br>
-          <a class="text-block__link" href="#">Need help?</a>
+          <a class="text-block__link" href="https://www.crossroads.net/forgot-password" target="_blank">Forgot password?</a><br>
```
Comment (sseissiger) [inline src/app/authentication/authentication.component.html]: Please build the urls from the current environment (int, demo, or prod)

File: src/app/services/payment.service.ts [hunk 8]
```diff
@@ -3,128 +3,124 @@ import { Http, Response } from '@angular/http';
...
+  makeApiDonorCall(donorInfo: CrdsDonor): Observable<any> {
+    let donorUrl = this.baseUrl + 'api/donor';
+    let requestOptions: any = this.httpClient.getRequestOption();
+
+    if (donorInfo.rest_method === this.restMethodNames.post) {
+      return this.http.post(donorUrl, donorInfo, requestOptions)
+        .map(this.extractData)
+        .catch(this.handleError);
+    } else if (donorInfo.rest_method === this.restMethodNames.put) {
+      return this.http.put(donorUrl, donorInfo, requestOptions)
```
Comment (vchtelmakh) [inline src/app/services/payment.service.ts]: Should use http client service. This is for refactor later, not keeping this from being merged.

**Pull request merged**